awscli/autocomplete/model.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Model based auto-completer.
+
+This provides autocompletion based on information found
+in the `service-2.json` files.
+
+"""
+from collections import namedtuple
+from awscli.autocomplete import indexer
+
+
+CLIArgument = namedtuple('CLIArgument', ['argname', 'type_name',
+                                         'command', 'parent', 'nargs'])
+
+
+class ModelIndex(object):
+    """Retrieve command/param names through querying an index.
+
+    This class provides methods for retrieving valid command
+    and parameter names given some context.  It's used by
+    the model based autocompleter.
+
+    """
+    def __init__(self, db_filename):
+        self._db_filename = db_filename
+        self._db_connection = None
+
+    def _get_db_connection(self):
+        if self._db_connection is None:
+            self._db_connection = indexer.DatabaseConnection(
+                self._db_filename)
+        return self._db_connection
+
+    def command_names(self, lineage):
+        """Return command names given a lineage.
+
+        This ``lineage`` is the same concept used in the
+        AWS CLI command classes, except that it explicitly
+        uses ``aws`` as the first parent.  This is the list of
+        parent commands for a given CLI command.  For example,
+        ``aws ec2 <here>`` has a lineage of ``['ec2']``.  The
```

Comment:
Yeah I'll update the doc example to be `['aws', 'ec2']` instead of `['ec2']`.

I think it's just a cleaner design that doesn't assume the "aws" prefix as parts of the value for lineage.  It makes the logic more generalized.  That way if you want to refer to the top level command you can use `lineage=[], command_name='aws'`.